src/Compilers/Core/Portable/Diagnostic/DiagnosticInfo.cs

```diff
@@ -122,6 +122,25 @@ internal static void AssertMessageSerializable(object[] args)
             }
         }
 
+        [Conditional("DEBUG")]
+        internal static void AssertExpectedMessageArgumentsLength(CommonMessageProvider messageProvider, int errorCode, int actualLength)
```

Comment:
Is this path for compiler diagnostics only?

I think it would be beneficial to have the assert for all diagnostics (to prevent IDExxxx from having bugs too)

Then the following analyzer will need to be excluded:

https://github.com/dotnet/roslyn/blob/8defa4afa5efbd0971a4bfcba7cc63fc8325a580/src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs#L894-L914

Not sure if that makes sense?

Reply:
Limiting the change to C# compiler diagnostics for now, to keep the scope manageable. We can consider addressing other diagnostics as needed.

Reply:
> I think it would be beneficial to have the assert for all diagnostics (to prevent IDExxxx from having bugs too)

I have no objection, as long as this is a debug-only assertion. It's possible we have runtime scenarios where the diagnostic intentionally provides more placeholders than are used by the message.